feat: 소셜로그인을 구현한다

drunkenhw opened the pull request:
## 📄 Summary
>  키아라가 만들어놓은 jwt 를 이용해서 소셜 로그인이 성공하면 토큰을 발급하도록 변경했습니다. 먼저 흐름은 
1. 프론트에서 원하는 벤더사의 이름과, redirect uri를 받아옵니다.
2. 서버에서는 받은 정보들로 해당 벤더사의 로그인 uri를 제공합니다
3. 프론트에서 로그인에 성공하면 벤더사에서 받아온 code와 redirect uri를 받습니다.
4. 받고 벤더사와 통신하여 정보를 받고 저장합니다
5. jwt를 리턴 해줍니다.

## 🕰️ Actual Time of Completion
>5시간

## 🙋🏻 More
> 


close #87

drunkenhw committed:
chore: application 의 property 읽는 의존성 추가

drunkenhw committed:
chore: application 의 property 읽는 의존성 추가

drunkenhw committed:
feat: OAuth2 회원 추가

drunkenhw committed:
feat: OAuth2 회원 추가

drunkenhw committed:
feat: provider 에게 회원 정보를 요청하는 기능 추가

drunkenhw committed:
feat: provider 에게 회원 정보를 요청하는 기능 추가

drunkenhw committed:
feat: 소셜 회원 정보를 조회하고 저장하는 기능 추가

drunkenhw committed:
feat: 소셜 회원 정보를 조회하고 저장하는 기능 추가

drunkenhw committed:
style: 컨벤션에 맞게 개행 추가 및 제거

[#87]

drunkenhw committed:
style: 컨벤션에 맞게 개행 추가 및 제거

[#87]

drunkenhw committed:
refactor: 로그인 시 jwt 발급하도록 변경

[#87]

drunkenhw committed:
fix: 에러나는 테스트 수정

[#87]

drunkenhw committed:
fix: 에러나는 테스트 수정

[#87]

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/auth/domain/TokenProvider.java:
```diff
@@ -1,10 +1,11 @@
     boolean isExpired(String token);
 
     Long extract(String token);
+
```
불필요한 개행은 없애줘도 될 것 같습니다.

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/auth/infrastructure/RestTemplateOAuthRequester.java:
```diff
@@ -0,0 +1,93 @@
+    @Override
+    public String loginUri(Provider provider, String redirectUri) {
+        OAuthProviderProperty providerProperties = oAuthProviderProperties.getProviderProperties(provider);
+        return UriComponentsBuilder.fromUriString(providerProperties.getLoginUrl())
```
`fromUriString`보다는 차라리 from은 어떨까요? String인 걸 알 필요가 있을까 싶어서요

drunkenhw replied:
해당 클래스는 다른 라이브러리라서 제가 어쩔 수 없었습니다

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/auth/infrastructure/RestTemplateOAuthRequester.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private OAuthTokenResponse requestAccessToken(OAuthProviderProperty property, OAuthLoginRequest loginRequest) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setBasicAuth(property.getId(), property.getSecret());
+        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
```
어떤 행위인지 메서드를 한 번 분리해줘도 좋을 것 같아요

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/auth/infrastructure/RestTemplateOAuthRequester.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private Map<String, Object> getUserAttributes(OAuthProviderProperty property, OAuthTokenResponse tokenResponse) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setBearerAuth(tokenResponse.accessToken());
+        headers.setContentType(MediaType.APPLICATION_JSON);
```
여기도 마찬가지입니다~

sosow0212 commented on backend/src/main/resources/application-prod.yml:
```diff
@@ -15,3 +15,5 @@ initialize-charge:
 
 scheduling:
   enabled: false
+
+
```
개행 없애줘도 될 것 같아요!

sosow0212 commented on backend/src/test/java/com/carffeine/carffeine/auth/domain/ProviderTest.java:
```diff
@@ -0,0 +1,37 @@
+        // when & then
+        assertThatThrownBy(() -> Provider.from(wrongName))
+                .isInstanceOf(AuthException.class)
+                .hasMessage("지원하지 않는 로그인 플랫폼입니다");
```
`Enum.getMessage()`로 메시지 검증하는 건 어떨까요?

sosow0212 commented on backend/src/test/java/com/carffeine/carffeine/auth/infra/FakeGoogleController.java:
```diff
@@ -0,0 +1,62 @@
+public class FakeGoogleController {
+
+    @PostMapping("/v4/token")
+    public OAuthTokenResponse token(
```
어떤 행위인지 구체적으로 네이밍해주면 더 좋을 것 같아요

sosow0212 commented on backend/src/test/java/com/carffeine/carffeine/auth/infra/FakeGoogleController.java:
```diff
@@ -0,0 +1,62 @@
+            @RequestParam(name = "redirect_uri") String redirectUri,
+            @RequestHeader(value = "Authorization", required = false) String authorizationHeader
+    ) {
+        if (isBasicAuthValid(authorizationHeader) &&
+                code.equals("carffeine") &&
+                grantType.equals("authorization_code") &&
+                redirectUri.equals("http://localhost:8080/")) {
```
이 부분도 메서드 분리를 통해서 어떤 행위인지 명시해주면 좋을 것 같습니다

sosow0212 commented on backend/src/test/java/com/carffeine/carffeine/auth/service/OAuthProviderPropertiesTest.java:
```diff
@@ -0,0 +1,33 @@
+import static com.carffeine.carffeine.auth.service.OAuthProviderProperties.OAuthProviderProperty;
+import static org.assertj.core.api.SoftAssertions.assertSoftly;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
```
👍

sosow0212 commented on backend/src/test/java/com/carffeine/carffeine/fake/member/FakeMemberRepository.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Optional;
+
+public class FakeMemberRepository implements MemberRepository {
+    private final Map<Long, Member> map = new HashMap<>();
```
개행 한 칸 추가해주세요

sosow0212 reviewed (changes requested):
작업하시느라 수고하셨습니다~ 기능적으로는 잘 해주신 것 같아서 컨벤션에 대해 몇 가지 남겼습니다.

drunkenhw committed:
refactor: 리뷰 반영

[#87]

drunkenhw committed:
refactor: 리뷰 반영

[#87]

be-student committed:
feat: docker 시작 전에 종료된 컨테이너를 제거한다 (#358)

sosow0212 committed:
refactor: context-path를 설정함으로 '/api'를 작성하지 않아도 prefix로 고정되도록 변경 (#383)

* refactor: context-path를 설정함으로 '/api'를 작성하지 않아도 prefix로 고정되도록 변경

* refactor: context-path를 설정함으로 '/api'를 작성하지 않아도 prefix로 고정되도록 변경

drunkenhw committed:
fix: order 순서 변경 (#386)

[#385]

gabrielyoon7 committed:
refactor: 폴더 구조와 파일 이름을 개선한다 (#387)

* refactor: Charger, Station 공통 타입 새로 추가

[#382]

* refactor: StationSummary, StationDetails 분리

[#382]

* refactor: 타입 분리 완료

[#382]

* rename: briefStationInfo를 stationSummary로 이름 변경

[#382]

* rename: briefStationInfo를 stationSummary로 이름 변경

[#382]

* refactor: 폴더 위치 수정

[#382]

* refactor: 폴더 위치 수정

[#382]

* refactor: 폴더 위치 수정

[#382]

* rename: 명확한 이름으로 개선

[#382]

* rename: 명확한 이름으로 개선

[#382]

* refactor: 필터 관련 store 위치 수정

[#382]

* refactor: 레이아웃 관련 스토어 파일 위치 수정

[#382]

Co-authored-by: kyw0716 <kyw0716@naver.com>
Co-authored-by: Dain Lee <feb.dain@gmail.com>

gabrielyoon7 committed:
fix: 버튼의 variant에 따라 크기가 달라지는 문제를 해결한다 (#390)

* fix: variant마다 다른 두께가 잘못 적용되는 문제 수정

[#389]

* test: 버튼 스위치 예제 추가

[#389]

be-student commented on backend/src/main/java/com/carffeine/carffeine/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,40 @@
+
+    private final AuthService authService;
+
+    @PostMapping("/oauth/{provider}/login-uri")
+    public ResponseEntity<LoginUriResponse> getRedirectUri(
+            @PathVariable String provider,
+            @RequestBody OAuthUriRequest request
+    ) {
+        String redirectUri = authService.loginUri(request, provider);
+        return ResponseEntity.ok(new LoginUriResponse(redirectUri));
+    }
```
이 요청이 post 인 이유가 있으신가요?

drunkenhw replied:
그러게요 수정했습니다~

be-student commented on backend/src/main/java/com/carffeine/carffeine/auth/domain/OAuthMember.java:
```diff
@@ -0,0 +1,12 @@
+
+    String id();
+
+    String imageUrl();
```
imageUrl 이 모든 provider 에 있는 정보인가요?

drunkenhw replied:
네 구글, 카카오, 네이버, 정도로 저희가 사용할 것 같은데 이 3개의 벤더사 그리고 깃헙에도 제공되는 것으로 알고있습니다.

drunkenhw committed:
feat: enum 컨버터 구현 (#392)

[#391]

kyw0716 committed:
refactor: 아코디언 컴포넌트를 개선한다 (#394)

* refactor: 중복된 네이밍 수정 Navigator → Menu

[#369]

* refactor: 아코디언 컴포넌트 수정

- isBasePanelOpen, isLastPanelOpen 등의 상태 삭제
- BasePanel, LastPanel에서 전역적으로 관리하고 있는 상태를 참조해 이를 렌더링 하는 방식으로 변경
- 이제는 좌측 메뉴 바에서 아이콘을 두번 눌러도 메뉴가 꺼지지 않음 (네이버가 이렇게 하기도 하고, 추가 로직 작성하면 너무 복잡해질거 같아서 이렇게 구현)

[#369]

* refactor: 닫기 버튼 활성화 props 추가

[#369]

* refactor: 스토리북 컴포넌트명 수정

[#369]

gabrielyoon7 committed:
refactor: 충전소 정보 제안 관련 기능을 개선한다 (#393)

* feat: 충전소 정보 불일치 신고 시 중복 제출이 불가능 하도록 로딩 효과 구현

[#378]

* feat: 충전소 정보 불일치 신고 시 100자 이상 입력이 불가능하도록 기능 구현

[#378]

* refactor: 불필요한 props 제거

[#378]

* refactor: 피드백 반영

- 글자 수 제한을 데이터베이스와 맞춤

[#378]

kiarakim committed:
refactor: 생성일과 수정일을 추가한다. (#374)

* feat: 생성일과 수정일을 상속할 부모 클래스 작성

* refactor: BaseTime 상속하도록 변경

* chore: prettier 3.0 버전으로 올리고 type 절대 경로를 추가한다 (#372)

* chore: `@trivago/prettier-plugin-sort-imports`, `prettier` 최신 버전으로 업데이트

[#371]

* chore: `@type` 절대 경로 추가

[#371]

* chore: 절대 경로로 변경

[#371]

* refactor: 클래스명 변경

* refactor: 와일드 import 제거

* refactor: jdbs 수정

* refactor: 수정일 호출하도록 변경

* refactor: @EnableJpaAuditing 제거

* refactor: 개행 추가

* refactor: Entity 생성 및 수정 시 시간 자동 저장하는 어노테이션 추가

* refactor: BaseEntity 패키지 변경

* refactor: Application 레벨에서 현재시간으로 지정

* refactor: import문 추가

---------

Co-authored-by: Dain Lee <108778921+feb-dain@users.noreply.github.com>

gabrielyoon7 committed:
feat: 로딩 스켈레톤 컴포넌트를 충전소 조회에 적용한다 (#396)

kiarakim commented on backend/src/main/java/com/carffeine/carffeine/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,40 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api")
```
우리 api 그거 빼기로 하지 않았나요?

drunkenhw replied:
그렇네요 제거하겠습니다

kiarakim commented on backend/src/main/java/com/carffeine/carffeine/auth/controller/AuthFilter.java:
```diff
@@ -44,26 +44,25 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
     private void sendUnauthorizedError(HttpServletResponse response, String errorMessage) throws IOException {
         ExceptionResponse exceptionResponse = new ExceptionResponse(HttpServletResponse.SC_UNAUTHORIZED, errorMessage);
         response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
-        response.setContentType("application/json");
+        response.setContentType(MediaType.APPLICATION_JSON_VALUE);
```
이 둘의 차이가 뭐에요??

drunkenhw replied:
Enum과 글자?

be-student commented on backend/src/main/java/com/carffeine/carffeine/auth/infrastructure/RestTemplateOAuthRequester.java:
```diff
@@ -0,0 +1,103 @@
+        HttpHeaders headers = headerWithToken(tokenResponse);
+        URI uri = URI.create(property.getInfoUrl());
+        RequestEntity<?> requestEntity = new RequestEntity<>(headers, HttpMethod.GET, uri);
+        ResponseEntity<Map<String, Object>> responseEntity = restTemplate.exchange(requestEntity, new ParameterizedTypeReference<>() {
+        });
+        return responseEntity.getBody();
```
getForObject 같은 방식으로 해도 좋을 것 같아요

kiarakim commented on backend/src/main/java/com/carffeine/carffeine/auth/controller/AuthMemberResolver.java:
```diff
@@ -16,27 +16,25 @@
     }
 
     @Override
-    public Object resolveArgument(
+    public Long resolveArgument(
```
tokenProvider.extract(token);
이 메서드는 Long을 반환하는데 확장성을 위해 반환타입을 Object로 하신건가요?
궁금궁금

drunkenhw replied:
Long을 반환하도록 변경한건데요..

be-student commented on backend/src/main/java/com/carffeine/carffeine/auth/service/AuthService.java:
```diff
@@ -0,0 +1,33 @@
+    private final MemberRepository memberRepository;
+    private final OAuthRequester oAuthRequester;
+
+    @Transactional
+    public String oAuthLogin(OAuthLoginRequest request, String provider) {
+        OAuthMember oAuthMember = oAuthRequester.login(request, provider);
+        Member member = memberRepository.findByEmail(oAuthMember.email())
+                .orElseGet(() -> memberRepository.save(new Member(oAuthMember.email())));
+        return tokenProvider.create(member.getId());
+    }
```
Transactional 의 범위가 살짝 큰 것 같아요
request 를 하는 부분을 transactional 의 범위에서 제거해야 될 것 같아요

be-student reviewed (approved):
이 부분만 수정해주시면 좋을 것 같아요 👍

kiarakim reviewed (approved):
어프룹할게요. 수고하셨어요 한가지만 수정해주세요~

gabrielyoon7 committed:
fix: 클라이언트 빌드 버그를 수정한다 (#400)

* feat: 충전소 간단 정보 카드 스켈레톤 구현 및 적용

[#388]

* test: 스켈레톤이 적용 된 충전소 간단 정보 카드 스토리북 작성

[#388]

* feat: 충전소 상세정보 조회 로딩화면 스켈레톤 적용

[#388]

* feat: 충전소 혼잡도 조회 로딩화면 스켈레톤 적용

[#388]

* feat: 충전소 혼잡도 조회 로딩화면 스켈레톤 적용

[#388]

* feat: 조회된 충전소가 없는 경우를 대응하는 기능 구현

[#388]

* resolve merge conflict

[#388]

* refactor: 불필요한 태그 제거

[#388]

* refactor: 버튼 간결 조절 및 hr 태그 제거

[#388]

* refactor: Array 메서드 수정

[#388]

* refactor: 재활용 가능한 스타일드 컴포넌트 추출

[#388]

* chore: dependencies 수정

[#399]

drunkenhw committed:
feat: method get으로 변경

[#87]

drunkenhw committed:
feat: method get으로 변경

[#87]

drunkenhw committed:
feat: transaction 분리

[#87]

drunkenhw committed:
feat: transaction 분리

[#87]

gabrielyoon7 committed:
chore: dependencies 수정 (#403)

[#402]

drunkenhw committed:
test: rest docs 테스트 작성

[#87]

drunkenhw committed:
test: rest docs 테스트 작성

[#87]

drunkenhw committed:
test: rest docs 테스트 작성

[#87)]

drunkenhw committed:
Merge remote-tracking branch 'origin/feat/87' into feat/87

# Conflicts:
#	backend/src/main/java/com/carffeine/carffeine/member/controller/MemberController.java

[#87]

sosow0212 reviewed (approved):
수고하셨습니다 !!

kyw0716 reviewed (approved):
굿샷
